File: scripts/banksy_cluster.py
#!/usr/bin/env python3
"""Run BANKSY spatial clustering on an AnnData .h5ad.

Invoked by `spacetravlr banksy --h5ad PATH` via isolated `uv run` (see src/banksy_cluster.rs).
"""

from __future__ import annotations

import argparse
import logging
import sys
from pathlib import Path

import anndata as ad
import numpy as np
import scanpy as sc

NBR_WEIGHT_DECAY_CHOICES = ("scaled_gaussian", "uniform", "reciprocal", "ranked")


def ensure_spatial_coordinates(
    adata: ad.AnnData,
    coord_key: str = "spatial",
    x_col: str | None = None,
    y_col: str | None = None,
) -> tuple[str, str, str]:
    if coord_key in adata.obsm:
        return x_col or "x", y_col or "y", coord_key

    candidates = [
        ("array_col", "array_row"),
        ("x", "y"),
        ("X", "Y"),
        ("col", "row"),
    ]
    if x_col and y_col:
        candidates.insert(0, (x_col, y_col))

    for x_name, y_name in candidates:
        if x_name in adata.obs.columns and y_name in adata.obs.columns:
            adata.obsm[coord_key] = np.column_stack(
                [
                    adata.obs[x_name].to_numpy(dtype=np.float64),
                    adata.obs[y_name].to_numpy(dtype=np.float64),
                ]
            )
            return x_name, y_name, coord_key

    raise ValueError(
        "Could not find spatial coordinates. Provide --coord-key for an existing obsm entry "
        f"or --x-col/--y-col for obs columns. Available obsm keys: {list(adata.obsm.keys())}"
    )


def preprocess_for_banksy(adata: ad.AnnData) -> None:
    sc.pp.normalize_total(adata, target_sum=1e4)
    sc.pp.log1p(adata)
    sc.pp.scale(adata, max_value=10)


def run_banksy(
    adata: ad.AnnData,
    *,
    lambda_: float,
    num_neighbours: int,
    nbr_weight_decay: str,
    max_m: int,
    resolution: float,
    num_nn: int,
    pca_dims: int,
    partition_seed: int,
    num_iterations: int,
    coord_key: str,
    x_col: str | None,
    y_col: str | None,
    cluster_key: str,
    preprocess: bool,
) -> ad.AnnData:
    from banksy.initialize_banksy import initialize_banksy
    from banksy.run_banksy import generate_banksy_matrix, pca_umap, run_Leiden_partition

    if nbr_weight_decay not in NBR_WEIGHT_DECAY_CHOICES:
        raise ValueError(
            f"Unsupported nbr_weight_decay={nbr_weight_decay!r}. "
            f"Choose one of {NBR_WEIGHT_DECAY_CHOICES}."
        )

    adata = adata.copy()
    x_name, y_name, coord_key = ensure_spatial_coordinates(
        adata, coord_key=coord_key, x_col=x_col, y_col=y_col
    )
    if preprocess:
        preprocess_for_banksy(adata)

    coord_keys = (x_name, y_name, coord_key)
    banksy_dict = initialize_banksy(
        adata,
        coord_keys=coord_keys,
        num_neighbours=num_neighbours,
        nbr_weight_decay=nbr_weight_decay,
        max_m=max_m,
        plt_edge_hist=False,
        plt_nbr_weights=False,
        plt_agf_angles=False,
        plt_theta=False,
    )
    banksy_dict, _ = generate_banksy_matrix(
        adata, banksy_dict, [lambda_], max_m=max_m, verbose=False
    )
    pca_umap(banksy_dict, pca_dims=[pca_dims], plt_remaining_var=False, add_umap=False)
    results_df, _ = run_Leiden_partition(
        banksy_dict=banksy_dict,
        resolutions=[resolution],
        num_nn=num_nn,
        num_iterations=num_iterations,
        partition_seed=partition_seed,
        match_labels=False,
    )
    labels = np.asarray(results_df.iloc[0]["labels"].dense, dtype=np.int32)
    adata.obs[cluster_key] = labels.astype(str)
    adata.obsm[f"X_banksy_pca_{pca_dims}"] = np.asarray(
        banksy_dict[nbr_weight_decay][lambda_]["adata"].obsm[f"reduced_pc_{pca_dims}"],
        dtype=np.float64,
    )
    return adata


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Run BANKSY spatial clustering on an AnnData .h5ad")
    parser.add_argument("--h5ad", required=True, help="Input .h5ad path")
    parser.add_argument(
        "-o",
        "--output",
        default=None,
        help="Output .h5ad path (default: <input_stem>_banksy.h5ad)",
    )
    parser.add_argument("--lambda", dest="lambda_", type=float, default=0.2)
    parser.add_argument("--num-neighbours", type=int, default=15)
    parser.add_argument(
        "--nbr-weight-decay",
        choices=NBR_WEIGHT_DECAY_CHOICES,
        default="scaled_gaussian",
    )
    parser.add_argument("--max-m", type=int, default=1)
    parser.add_argument("--resolution", type=float, default=0.6)
    parser.add_argument("--num-nn", type=int, default=50)
    parser.add_argument("--pca-dims", type=int, default=20)
    parser.add_argument("--partition-seed", type=int, default=1234)
    parser.add_argument("--num-iterations", type=int, default=-1)
    parser.add_argument("--coord-key", default="spatial")
    parser.add_argument("--x-col", default=None)
    parser.add_argument("--y-col", default=None)
    parser.add_argument("--cluster-key", default="banksy_cluster")
    parser.add_argument("--no-preprocess", action="store_true")
    parser.add_argument("-v", "--verbose", action="store_true")
    args = parser.parse_args(argv)

    logging.basicConfig(
        level=logging.DEBUG if args.verbose else logging.INFO,
        format="%(levelname)s: %(message)s",
    )

    h5ad_path = Path(args.h5ad)
    output_path = (
        Path(args.output)
        if args.output
        else h5ad_path.with_name(f"{h5ad_path.stem}_banksy.h5ad")
    )

    logging.info("Loading %s", h5ad_path)
    adata = ad.read_h5ad(h5ad_path)
    result = run_banksy(
        adata,
        lambda_=args.lambda_,
        num_neighbours=args.num_neighbours,
        nbr_weight_decay=args.nbr_weight_decay,
        max_m=args.max_m,
        resolution=args.resolution,
        num_nn=args.num_nn,
        pca_dims=args.pca_dims,
        partition_seed=args.partition_seed,
        num_iterations=args.num_iterations,
        coord_key=args.coord_key,
        x_col=args.x_col,
        y_col=args.y_col,
        cluster_key=args.cluster_key,
        preprocess=not args.no_preprocess,
    )
    output_path.parent.mkdir(parents=True, exist_ok=True)
    result.obs.index = result.obs.index.astype(str)
    result.var.index = result.var.index.astype(str)
    ad.settings.allow_write_nullable_strings = True
    result.write_h5ad(output_path)
    n_clusters = result.obs[args.cluster_key].nunique()
    logging.info(
        "BANKSY finished: %d cells, %d clusters (lambda=%s, resolution=%s)",
        result.n_obs,
        n_clusters,
        args.lambda_,
        args.resolution,
    )
    print(f"wrote {output_path}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/banksy_cluster.rs
//! BANKSY spatial clustering: runs [`scripts/banksy_cluster.py`] in an isolated **`uv`** env.

use crate::config::resolve_banksy_cluster_py_path;
use crate::scanpy_preprocess::uv_python_stdin;
use anyhow::Context;
use std::fs;
use std::path::Path;

const BANKSY_SCRIPT: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/scripts/banksy_cluster.py"
));

pub const UV_WITH_BANKSY: &[&str] = &[
    "numpy<2",
    "pandas",
    "anndata>=0.10",
    "scipy",
    "scanpy",
    "h5py",
    "scikit-learn",
    "leidenalg",
    "igraph",
    "python-igraph",
    "pybanksy>=1.3.5",
];

pub struct BanksyParams<'a> {
    pub h5ad: &'a Path,
    pub output: Option<&'a Path>,
    pub lambda: f64,
    pub num_neighbours: u32,
    pub nbr_weight_decay: &'a str,
    pub max_m: u32,
    pub resolution: f64,
    pub num_nn: u32,
    pub pca_dims: u32,
    pub partition_seed: u32,
    pub num_iterations: i32,
    pub coord_key: &'a str,
    pub x_col: Option<&'a str>,
    pub y_col: Option<&'a str>,
    pub cluster_key: &'a str,
    pub preprocess: bool,
    pub verbose: bool,
}

pub fn run_banksy(params: BanksyParams<'_>) -> anyhow::Result<()> {
    let h5ad = params
        .h5ad
        .to_str()
        .with_context(|| format!("h5ad path must be UTF-8: {}", params.h5ad.display()))?;

    let mut argv: Vec<String> = vec![
        "--h5ad".into(),
        h5ad.into(),
        "--lambda".into(),
        params.lambda.to_string(),
        "--num-neighbours".into(),
        params.num_neighbours.to_string(),
        "--nbr-weight-decay".into(),
        params.nbr_weight_decay.into(),
        "--max-m".into(),
        params.max_m.to_string(),
        "--resolution".into(),
        params.resolution.to_string(),
        "--num-nn".into(),
        params.num_nn.to_string(),
        "--pca-dims".into(),
        params.pca_dims.to_string(),
        "--partition-seed".into(),
        params.partition_seed.to_string(),
        "--num-iterations".into(),
        params.num_iterations.to_string(),
        "--coord-key".into(),
        params.coord_key.into(),
        "--cluster-key".into(),
        params.cluster_key.into(),
    ];

    if let Some(out) = params.output {
        let s = out
            .to_str()
            .with_context(|| format!("output path must be UTF-8: {}", out.display()))?;
        argv.push("-o".into());
        argv.push(s.into());
    }
    if let Some(x) = params.x_col {
        if !x.trim().is_empty() {
            argv.push("--x-col".into());
            argv.push(x.into());
        }
    }
    if let Some(y) = params.y_col {
        if !y.trim().is_empty() {
            argv.push("--y-col".into());
            argv.push(y.into());
        }
    }
    if !params.preprocess {
        argv.push("--no-preprocess".into());
    }
    if params.verbose {
        argv.push("-v".into());
    }

    let argv_refs: Vec<&str> = argv.iter().map(|s| s.as_str()).collect();
    let script = if let Some(p) = resolve_banksy_cluster_py_path() {
        fs::read_to_string(&p).with_context(|| format!("read BANKSY script {}", p.display()))?
    } else {
        BANKSY_SCRIPT.to_string()
    };
    uv_python_stdin(
        UV_WITH_BANKSY,
        &script,
        &argv_refs,
        false,
        "banksy clustering",
    )?;
    Ok(())
}
